File: java/src/obsolete/Treap.java
package obsolete;
import java.util.*;

public class Treap {

	static class Node {
		int cnt;
		int prio;
		int value;
		Node l;
		Node r;
	}

	static Random rnd = new Random(1);
	Node root;

	public int size() {
		return cnt(root);
	}

	int cnt(Node p) {
		return p == null ? 0 : p.cnt;
	}

	void update(Node p) {
		if (p != null)
			p.cnt = 1 + cnt(p.l) + cnt(p.r);
	}

	Node[] split(Node t, int index) {
		if (t == null)
			return new Node[2];
		Node[] res;
		if (index <= cnt(t.l)) {
			Node[] r = split(t.l, index);
			t.l = r[1];
			res = new Node[] { r[0], t };
		} else {
			Node[] r = split(t.r, index - cnt(t.l) - 1);
			t.r = r[0];
			res = new Node[] { t, r[1] };
		}
		update(t);
		return res;
	}

	Node merge(Node l, Node r) {
		Node res;
		if (l == null)
			res = r;
		else if (r == null)
			res = l;
		else if (l.prio > r.prio) {
			l.r = merge(l.r, r);
			res = l;
		} else {
			r.l = merge(l, r.l);
			res = r;
		}
		update(res);
		return res;
	}

	int get(Node t, int index) {
		if (index < cnt(t.l))
			return get(t.l, index);
		else if (index > cnt(t.l))
			return get(t.r, index - cnt(t.l) - 1);
		return t.value;
	}

	public int get(int index) {
		return get(root, index);
	}

	Node add(Node t, Node it, int index) {
		Node res;
		if (t == null)
			res = it;
		else if (it.prio < t.prio) {
			Node[] r = split(t, index);
			it.l = r[0];
			it.r = r[1];
			res = it;
		} else {
			if (index <= cnt(t.l))
				t.l = add(t.l, it, index);
			else
				t.r = add(t.r, it, index - cnt(t.l) - 1);
			res = t;
		}
		update(res);
		return res;
	}

	public void add(int index, int value) {
		Node it = new Node();
		it.value = value;
		it.prio = rnd.nextInt();
		root = add(root, it, index);
	}

	Node remove(Node t, int index) {
		Node res;
		if (index == cnt(t.l))
			res = merge(t.l, t.r);
		else {
			if (index < cnt(t.l))
				t.l = remove(t.l, index);
			else
				t.r = remove(t.r, index - cnt(t.l) - 1);
			res = t;
		}
		update(res);
		return res;
	}

	public void remove(int index) {
		root = remove(root, index);
	}

	void print(Node t) {
		if (t != null) {
			print(t.l);
			System.out.print(t.value + " ");
			print(t.r);
		}
	}

	public void print() {
		print(root);
		System.out.println();
	}

	// Usage example
	public static void main(String[] args) {
		Treap treap = new Treap();
		treap.add(0, 1);
		treap.add(0, 2);
		treap.add(0, 3);
		treap.remove(2);
		treap.remove(1);
		for (int i = 0; i < treap.size(); i++) {
			System.out.println(treap.get(i));
		}
		treap.print();
		treap.remove(0);
	}
}

File: kotlin/Treap.kt
import java.util.Random
import java.util.TreeSet

// https://en.wikipedia.org/wiki/Treap
object Treap {
    val random = Random()

    data class Treap(
            val key: Int,
            var left: Treap? = null,
            var right: Treap? = null,
            var size: Int = 1,
            val prio: Long = random.nextLong()
    ) {
        fun update() {
            size = 1 + getSize(left) + getSize(right)
        }
    }

    fun getSize(root: Treap?) = root?.size ?: 0

    fun split(root: Treap?, minRight: Int): Pair<Treap?, Treap?> {
        if (root == null)
            return Pair(null, null)
        if (root.key >= minRight) {
            val leftSplit = split(root.left, minRight)
            root.left = leftSplit.second
            root.update()
            return Pair(leftSplit.first, root)
        } else {
            val rightSplit = split(root.right, minRight)
            root.right = rightSplit.first
            root.update()
            return Pair(root, rightSplit.second)
        }
    }

    fun merge(left: Treap?, right: Treap?): Treap? {
        if (left == null)
            return right
        if (right == null)
            return left
        if (left.prio > right.prio) {
            left.right = merge(left.right, right)
            left.update()
            return left
        } else {
            right.left = merge(left, right.left)
            right.update()
            return right
        }
    }

    fun insert(root: Treap?, key: Int): Treap? {
        val t = split(root, key)
        return merge(merge(t.first, Treap(key)), t.second)
    }

    fun remove(root: Treap?, key: Int): Treap? {
        val t = split(root, key)
        return merge(t.first, split(t.second, key + 1).second)
    }

    fun kth(root: Treap, k: Int): Int {
        if (k < getSize(root.left))
            return kth(root.left!!, k)
        else if (k > getSize(root.left))
            return kth(root.right!!, k - getSize(root.left) - 1)
        return root.key
    }

    fun print(root: Treap?) {
        if (root == null)
            return
        print(root.left)
        println(root.key)
        print(root.right)
    }

    // random test
    @JvmStatic
    fun main(args: Array<String>) {
        val time = System.currentTimeMillis()
        var treap: Treap? = null
        val set = TreeSet<Int>()
        for (i in 0..999999) {
            val key = random.nextInt(100000)
            if (random.nextBoolean()) {
                treap = remove(treap, key)
                set.remove(key)
            } else if (!set.contains(key)) {
                treap = insert(treap, key)
                set.add(key)
            }
            if (set.size != getSize(treap))
                throw RuntimeException()
        }
        for (i in 0 until getSize(treap)) {
            if (!set.contains(kth(treap!!, i)))
                throw RuntimeException()
        }
        println(System.currentTimeMillis() - time)
    }
}
